File: ui/src/types/websocket.ts
/**
 * WebSocket message types and interfaces for real-time updates
 */

// Message types from server
export type MessageType =
  | 'training_progress'
  | 'backtest_progress'
  | 'log_entry'
  | 'system_metrics'
  | 'job_complete'
  | 'job_error'
  | 'pong'
  | 'subscribed'
  | 'unsubscribed'
  | 'error';

// Message types to server
export type ClientMessageType = 'subscribe' | 'unsubscribe' | 'ping';

// Subscription channels
export type Channel = 'training' | 'backtest' | 'logs' | 'system';

// Client message structure
export interface ClientMessage {
  type: ClientMessageType;
  channel?: Channel;
  jobId?: string;
}

// Server message structure
export interface ServerMessage<T = unknown> {
  type: MessageType;
  data: T;
  timestamp: string;
}

// Training progress data
export interface TrainingProgressData {
  jobId: string;
  status: string;
  progress: {
    currentEpoch?: number;
    totalEpochs?: number;
    percent?: number;
    currentTimestep?: number;
    totalTimesteps?: number;
  };
  phase?: 'transformer' | 'agent';
  epoch?: number;
  totalEpochs?: number;
  trainLoss?: number;
  valLoss?: number;
  learningRate?: number;
  elapsedSeconds?: number;
  estimatedRemainingSeconds?: number;
}

// Backtest progress data
export interface BacktestProgressData {
  jobId: string;
  status: string;
  currentBar?: number;
  totalBars?: number;
  currentEquity?: number;
  currentDrawdown?: number;
  tradesExecuted?: number;
}

// Log entry data
export interface LogEntryData {
  timestamp: string;
  level: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR';
  logger: string;
  message: string;
}

// System metrics data
export interface SystemMetricsData {
  cpu: {
    percent: number;
    cores: number;
  };
  memory: {
    total: number;
    used: number;
    percent: number;
  };
  gpu?: {
    available: boolean;
    name?: string;
    memory?: {
      total: number;
      used: number;
      percent: number;
    };
  };
  disk: {
    total: number;
    used: number;
    percent: number;
  };
}

// Job complete data
export interface JobCompleteData {
  jobId: string;
  status: 'completed';
  result?: {
    finalLoss?: number;
    bestEpoch?: number;
    modelPath?: string;
  };
}

// Job error data
export interface JobErrorData {
  jobId: string;
  error: {
    code?: string;
    message: string;
    recoverable: boolean;
  };
}

// Subscription confirmation data
export interface SubscriptionData {
  channel: Channel;
  jobId?: string;
}

// Error data
export interface ErrorData {
  message: string;
}

// WebSocket connection status
export type ConnectionStatus = 'connecting' | 'connected' | 'disconnected' | 'reconnecting';

// WebSocket hook return type
export interface UseWebSocketReturn {
  status: ConnectionStatus;
  subscribe: (channel: Channel, jobId?: string) => void;
  unsubscribe: (channel: Channel, jobId?: string) => void;
  lastMessage: ServerMessage | null;
  trainingProgress: Map<string, TrainingProgressData>;
  backtestProgress: Map<string, BacktestProgressData>;
  systemMetrics: SystemMetricsData | null;
  logs: LogEntryData[];
}

File: api/app/routes/websocket.py
"""WebSocket route for real-time updates."""

import logging
from datetime import datetime, timezone

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from ..core.job_manager import Job, job_manager
from ..core.websocket_manager import Channel, MessageType, ws_manager

logger = logging.getLogger(__name__)

router = APIRouter(tags=["websocket"])


async def job_update_callback(job: Job) -> None:
    """Callback for job updates - broadcasts to WebSocket subscribers."""
    # Determine channel and message type based on job type
    channel = Channel.TRAINING.value if "training" in job.job_type.value else Channel.BACKTEST.value

    # Build progress data
    data = {
        "jobId": job.job_id,
        "status": job.status.value,
        "progress": job.progress,
    }

    # Add job-specific data
    if job.job_type.value == "training":
        data.update({
            "phase": job.progress.get("phase", "transformer"),
            "epoch": job.progress.get("currentEpoch"),
            "totalEpochs": job.progress.get("totalEpochs"),
            "trainLoss": job.progress.get("trainLoss"),
            "valLoss": job.progress.get("valLoss"),
            "learningRate": job.progress.get("learningRate"),
        })
        if job.started_at:
            elapsed = (job.completed_at or datetime.now(timezone.utc)) - job.started_at
            data["elapsedSeconds"] = int(elapsed.total_seconds())
            # Estimate remaining time based on epoch progress
            if job.progress.get("currentEpoch") and job.progress.get("totalEpochs"):
                current = job.progress["currentEpoch"]
                total = job.progress["totalEpochs"]
                if current > 0:
                    time_per_epoch = elapsed.total_seconds() / current
                    remaining_epochs = total - current
                    data["estimatedRemainingSeconds"] = int(time_per_epoch * remaining_epochs)
    else:
        data.update({
            "currentBar": job.progress.get("currentBar"),
            "totalBars": job.progress.get("totalBars"),
            "currentEquity": job.progress.get("currentEquity"),
            "currentDrawdown": job.progress.get("currentDrawdown"),
            "tradesExecuted": job.progress.get("tradesExecuted"),
        })

    # Determine message type based on status
    if job.status.value == "completed":
        msg_type = MessageType.JOB_COMPLETE
        data["result"] = job.result
    elif job.status.value == "failed":
        msg_type = MessageType.JOB_ERROR
        data["error"] = {"message": job.error, "recoverable": False}
    else:
        msg_type = (
            MessageType.TRAINING_PROGRESS
            if channel == Channel.TRAINING.value
            else MessageType.BACKTEST_PROGRESS
        )

    # Broadcast to subscribers
    await ws_manager.broadcast_to_channel(channel, msg_type, data, job.job_id)


@router.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    """WebSocket endpoint for real-time updates."""
    await ws_manager.connect(websocket)

    try:
        while True:
            # Receive and parse message
            try:
                data = await websocket.receive_json()
                await ws_manager.handle_message(websocket, data)
            except json.JSONDecodeError:
                await ws_manager.send_personal_message(
                    websocket,
                    MessageType.ERROR,
                    {"message": "Invalid JSON"},
                )

    except WebSocketDisconnect:
        logger.info(f"WebSocket disconnected normally")
    except Exception as e:
        logger.error(f"WebSocket error: {e}")
    finally:
        await ws_manager.disconnect(websocket)


def setup_job_subscribers():
    """Set up subscribers for existing and new jobs."""
    # This function should be called on startup to wire up job notifications
    # Subscribe to all existing jobs
    for job_id, job in job_manager.jobs.items():
        job.subscribe(job_update_callback)


# Monkey-patch job creation to auto-subscribe new jobs
_original_create_job = job_manager.create_job


def _create_job_with_subscriber(job_type, config):
    """Wrapper to add WebSocket subscriber to new jobs."""
    job = _original_create_job(job_type, config)
    job.subscribe(job_update_callback)
    return job


job_manager.create_job = _create_job_with_subscriber
